### ISSUE TITLE
Simplify the syntax for Setup and Verifying with some new syntactic sugar

### ISSUE THREAD

**Issue by PureKrome** [FakeItEasy/FakeItEasy#539]
Hi team,

:heart: FiE :) Been using it for ages :)

Just thought I might try a weee little discussion about something which myself and a few other peeps sorta have a tiny gripe re: FiE. For the record, this discussion is based upon our experience from using moq, ages ago before switching.

- The syntax for setup and verify can sometimes be a little awkward. 
- The verify (`MustHave..`, `MustNotHave..` .. etc) is so verbose.

---
1. Setup/Verify.

So to setup something we use `A.CallTo( () => ..)` and for verify, we use that, again. Sure, I've been using FiE for ages so I get it, but it can still be confusing when doing code reviews because at a quick glance I don't always see what it's trying to do.

Is there any thoughts into adding some syntactic sugar for `Setup( () => userRepo.Foo(....))` which will not compile if you add `MustNotHave...` at the end? So it's explicitly a setup. (or whatever word you use to use... like `Setup` or `When` etc..).

---

2. Verify is too verbose.

compare:

`A.CallTo( () => userRepo.GetUser(1)).MustHaveHappened(Repeated.Exactly.Once));`

vs

`A.CallTo( () => userRepo.GetUser(1)).MustHaveHappened(ExactlyOnce);`

(or even better)

`Verify( () => userRepo.GetUser(1)).HappenedExactlyOnce;`

Most of the time we use
- MustNotHaveHappend();
- Exactly once
- More than once.

  ... and sure .. the above change is small.

----

Ok, so I grok that I can add my own syntactic sugar etc. But I use this library _everywhere_ and instead of making some tiny library that I end up using only, I'm hoping that other people might agree with my thoughts - including the repo owners ... and maybe add these _options_ to the library. 

Thoughts?

**Comment by blairconrad** [comment]
@PureKrome, we'r'e glad you enjoy FakeItEasy. Thanks!
Most of the time when I see requests of this sort, I'm initially against them. Just as I am with any features.
So I took a few days to go away and think about things and come back and reread this issue.

I wish I could say that my feelings have changed from my initial reaction, but they haven't.

Regarding shortening the verify, I don't even know how it would be possible to make the syntax
`A.CallTo( () => userRepo.GetUser(1)).MustHaveHappened(ExactlyOnce);`. What is `ExactlyOnce` in here? And even if it were possible, a savings of "Repeated.." doesn't really seem worthwhile when factoring in the cost of adding a new concept for users.

As far as the `Setup`/`Verify` syntactic sugar goes, I appreciate the idea that the `Setup` would not support a `Must(Not)HaveHappened` on end of it. That does seem beneficial. But it's not a mistake that I see cropping up frequently. You mention the confusion during code reviews. Is this because the developer is using it incorrectly, or because the reviewer is missing it? Neither is great, but I'd worry more about the former.

Part of FakeItEasy's philosophy is to keep the surface area small: not to add to many concepts to the API and above all to make them easily discoverable. As much as possible is supposed to be driven from the `A` class.  And we explicitly tout the fact that the configure/verify syntax is the same:

```c#
// Asserting uses the same syntax as configuring calls.
// There's no need to learn another syntax.
A.CallTo(() => shop.BuyCandy(lollipop)).MustHaveHappened();
```

(nicked right from the [web site ](http://fakeiteasy.github.io/)). 

I realize that `Setup` and `Verify` could be added as options, but I just don't see it being worth the cost involved in developing, testing, documenting, and maintaining the extra entry points.

You are against making a tiny library that only you end up using, but probably most things on GitHub were tiny libraries that only the author used. If you think only you would use it, why do you think the additions to FakeItEasy would catch on? Perhaps they'd languish in obscurity as well.

Why not try making a NuGet package? It could depend on FakeItEasy, so you wouldn't even need to increase the net total of top-level packages your projects depend on. Then you can develop the ideas. You'll get something you can share across your projects and it would be a testbed for the sugar. If it provides sufficient value, perhaps others will flock to it, and the improvements will be seen to be worth rolling into FakeItEasy, bringing along an eager audience as well as a reference implementation.

I'm not trying to crush your dreams, @PureKrome. It's just that I don't see the value relative the anticipated audience size and the work required to enact the changes.

@FakeItEasy/owners, unless something else comes up, I'd be inclined to close **wontfix**.

As always, I'm happy to hear differing opinions from anyone at all.

**Comment by PureKrome** [comment]
>Regarding shortening the verify, I don't even know how it would be possible to make the syntax
A.CallTo( () => userRepo.GetUser(1)).MustHaveHappened(ExactlyOnce);. What is ExactlyOnce in here? 

It's the exact equiv of `MustHaveHappened(Repeated.Exactly.Once);`

I wasn't suggesting taking away the above line/syntax, just offering a shorter and simpler version to the same thing. Especially when it comes to intellisence and quick typing. For example, I would type

- `dot`
- `M`
- `H`
- `H`
- `E`
- `O`
- enter 

.. and that autocompletes the full text `.MustHaveHappenedExactlyOnce`. 

> As far as the Setup/Verify syntactic sugar goes, I appreciate the idea that the Setup would not support a Must(Not)HaveHappened on end of it. That does seem beneficial. But it's not a mistake that I see cropping up frequently.

Completely agree -> the M(N)HH only ever occurs when people are verifying. Either that or `Returns`.

>You mention the confusion during code reviews. Is this because the developer is using it incorrectly, or because the reviewer is missing it?

More the second. The developer has two choices -> `Returns` or `M(N)HH`. What I was suggesting was that during our code reviews or just code support/maintenance, we need to read the ENTIRE line before we can grok what they are trying to do. For example, if we are in the `Arrange` part of an `Arrange/Act/Assert` test, there's lots of `A.Call( () => ..)` methods and sure... you tell yourself that this is the `Arrange` section so it must be doing setups. But on a quick glance it's not comforting nor 100% confident. It is possible that they might have mixed things up or just done something wrong (eg. cut-n-paste baddies).

The point was that on a first 1 sec visual glance it's not 100% completely obvious IMO.

Think of this like an interlaced picture. The first pass is -really- quick and nasty. But more passes/scans will definitely happen in a second later. But during that first pass/scan as my eyes pass over the code, i'm trying to be efficient and segment things into quick summaries in my head. And for me, having some code syntax that could go either way (setup or verify) to me, slows things down.

>Part of FakeItEasy's philosophy is to keep the surface area small: not to add to many concepts to the API and above all to make them easily discoverable. 

Agree :) :+1: which is also part of why I found it easy to start using the library.

>As much as possible is supposed to be driven from the A class. And we explicitly tout the fact that the configure/verify syntax is the same

I'll put this down to my personal opinion which is that I find this part of the API confusing. sure, it's a single api call ... but it's now doing two seperate things. I like to separate my concerns. I see the concerns outweight the discovery. Again -> just my personal opinion and i'm not hating, etc.

>You are against making a tiny library that only you end up using, but probably most things on GitHub were tiny libraries that only the author used. If you think only you would use it, why do you think the additions to FakeItEasy would catch on? Perhaps they'd languish in obscurity as well.

My suggestion there was to leverage the spotlight of FiE. Sure, I can go off and make my own extensions to this library. Then NuGet that bad boy. But then I stuck with an extension which I feel might be valuable but would never get traction, cause seriously, i'm just another statistic in the massive site. So to see if my idea could slip into the main repo could mean more people might find that valuable and use it. Secondly, it's also some validation about my thoughts. Lastly, it's also 1 less nuget package i have to pull down into my repo's. Sure it's nothing and u can have heaps of nuget packages in your project but having one less just makes me feel a little bit less clutered and a (false?) sense of less clutter and complexity == slightly easier to maintain. yet again, another personal opinion/mind set.

>I don't see the value relative the anticipated audience size and the work required to enact the changes.

Yep - fair enough. Without asking, I would never know :)

**Comment by blairconrad** [comment]
> >Regarding shortening the verify, I don't even know how it would be possible to make the syntax
`A.CallTo( () => userRepo.GetUser(1)).MustHaveHappened(ExactlyOnce);`. What is ExactlyOnce in here? 

> It's the exact equiv of `MustHaveHappened(Repeated.Exactly.Once);`

But `Repeated.Exactly.Once` is a member on a static member on the
`Repeated` class. I'm not sure how to make an `ExactlyOnce` that will
satisfy the compiler. But maybe I shouldn't argue implementation when
I'm still not on board with the overall idea...

> I wasn't suggesting taking away the above line/syntax, just offering
> a shorter and simpler version to the same thing.

Understood.

> Especially when it comes to intellisence and quick typing. For example, I would type
> 
> - `dot`
> - `M`
> - `H`
> - `H`
> - `E`
> - `O`
> - enter 
> 
> .. and that autocompletes the full text `.MustHaveHappenedExactlyOnce`. 

May I suggest a plugin that supports snippets or live templates or
something? Maybe even AutoHotKey. :grinning:

> >As much as possible is supposed to be driven from the A class. And we explicitly tout the fact that the configure/verify syntax is the same

> I'll put this down to my personal opinion which is that I find this
> part of the API confusing. sure, it's a single api call ... but it's
> now doing two seperate things. I like to separate my concerns. I see
> the concerns outweight the discovery. Again -> just my personal
> opinion and i'm not hating, etc.

Understood. I know you've good intentions! 

I'd argue that the `A.CallTo` API is doing a single thing -
identifying a call. What anyone decides to do with that is their own
business. Feels like arguing to me that `object.ToString()` is doing
two separate things because you can write
`obj.ToString().Replace("hippo", "lemur")` and
`obj.ToString().IndexOf("giraffe")`. `ToString` is still just making a
string representation of an object, even though you could use the
result for replacing _or_ searching.


> >You are against making a tiny library that only you end up using, but probably most things on GitHub were tiny libraries that only the author used. If you think only you would use it, why do you think the additions to FakeItEasy would catch on? Perhaps they'd languish in obscurity as well.

> My suggestion there was to leverage the spotlight of FiE. Sure, I can go off and make my own extensions to this library. Then NuGet that bad boy. But then I stuck with an extension which I feel might be valuable but would never get traction, cause seriously, i'm just another statistic in the massive site.

But at least it would be a statistic. You'd get some information about usage. Adding the methods to FakeItEasy means we'd have no insight into usage.

> Lastly, it's also 1 less nuget package i have to pull down into my repo's. Sure it's nothing and u can have heaps of nuget packages in your project but having one less just makes me feel a little bit less clutered and a (false?) sense of less clutter and complexity == slightly easier to maintain. yet again, another personal opinion/mind set.

Agreed about opinion/mindset. The extra package wouldn't bother _me_ one bit, if I found value in it. You could always ILMerge FakeItEasy! Okay. Maybe that's ridiculous.

> >I don't see the value relative the anticipated audience size and the work required to enact the changes.

> Yep - fair enough. Without asking, I would never know :)

True. And I could be completely wrong. It's all speculation.

**Comment by PureKrome** [comment]
Thanks for your time Blair.

**Comment by blairconrad** [comment]
Thank you, @PureKrome! I appreciate your interest, and the ideas, which I think could be useful, but perhaps not as broadly as we'd need for inclusion.
Believe it or not, I don't _enjoy_ being "Dr. No".
